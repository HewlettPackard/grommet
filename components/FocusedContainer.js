"use strict";

exports.__esModule = true;
exports.FocusedContainer = void 0;
var _react = _interopRequireWildcard(require("react"));
var _utils = require("../utils");
var _RootsContext = require("../contexts/RootsContext");
var _excluded = ["hidden", "restrictScroll", "children", "trapFocus"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.indexOf(n) >= 0) continue; t[n] = r[n]; } return t; }
var FocusedContainer = exports.FocusedContainer = function FocusedContainer(_ref) {
  var _ref$hidden = _ref.hidden,
    hidden = _ref$hidden === void 0 ? false : _ref$hidden,
    _ref$restrictScroll = _ref.restrictScroll,
    restrictScroll = _ref$restrictScroll === void 0 ? false : _ref$restrictScroll,
    children = _ref.children,
    trapFocus = _ref.trapFocus,
    rest = _objectWithoutPropertiesLoose(_ref, _excluded);
  var _useState = (0, _react.useState)(''),
    bodyOverflowStyle = _useState[0],
    setBodyOverflowStyle = _useState[1];
  var ref = (0, _react.useRef)(null);
  var roots = (0, _react.useContext)(_RootsContext.RootsContext);
  var _useState2 = (0, _react.useState)(roots),
    nextRoots = _useState2[0],
    setNextRoots = _useState2[1];
  (0, _react.useEffect)(function () {
    // make sure value of null is not added to array
    if (ref.current) setNextRoots([].concat(roots, [ref.current]));
  }, [roots]);
  (0, _react.useEffect)(function () {
    if (bodyOverflowStyle !== 'hidden' && !hidden && restrictScroll && trapFocus) {
      setBodyOverflowStyle(document.body.style.overflow);
      document.body.style.overflow = 'hidden';
    }
    return function () {
      if (bodyOverflowStyle !== 'hidden' && !hidden && restrictScroll && trapFocus) {
        document.body.style.overflow = bodyOverflowStyle;
      }
    };
  }, [bodyOverflowStyle, hidden, trapFocus, restrictScroll]);
  (0, _react.useEffect)(function () {
    var timer = setTimeout(function () {
      if (!hidden && trapFocus && roots && roots[0]) {
        roots.forEach(_utils.makeNodeUnfocusable);
      }
    }, 0);
    return function () {
      // remove trap and restore ability to focus on the last root only
      if (roots && roots[0]) (0, _utils.makeNodeFocusable)(roots[roots.length - 1]);
      clearTimeout(timer);
    };
  }, [hidden, roots, trapFocus]);
  return /*#__PURE__*/_react["default"].createElement(_RootsContext.RootsContext.Provider, {
    value: nextRoots
  }, /*#__PURE__*/_react["default"].createElement("div", _extends({
    ref: ref,
    "aria-hidden": hidden
  }, rest), children));
};