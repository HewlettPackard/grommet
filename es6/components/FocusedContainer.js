var _excluded = ["hidden", "restrictScroll", "children", "trapFocus"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.indexOf(n) >= 0) continue; t[n] = r[n]; } return t; }
import React, { useContext, useEffect, useRef, useState } from 'react';
import { makeNodeFocusable, makeNodeUnfocusable } from '../utils';
import { RootsContext } from '../contexts/RootsContext';
export var FocusedContainer = function FocusedContainer(_ref) {
  var _ref$hidden = _ref.hidden,
    hidden = _ref$hidden === void 0 ? false : _ref$hidden,
    _ref$restrictScroll = _ref.restrictScroll,
    restrictScroll = _ref$restrictScroll === void 0 ? false : _ref$restrictScroll,
    children = _ref.children,
    trapFocus = _ref.trapFocus,
    rest = _objectWithoutPropertiesLoose(_ref, _excluded);
  var _useState = useState(''),
    bodyOverflowStyle = _useState[0],
    setBodyOverflowStyle = _useState[1];
  var ref = useRef(null);
  var roots = useContext(RootsContext);
  var _useState2 = useState(roots),
    nextRoots = _useState2[0],
    setNextRoots = _useState2[1];
  useEffect(function () {
    // make sure value of null is not added to array
    if (ref.current) setNextRoots([].concat(roots, [ref.current]));
  }, [roots]);
  useEffect(function () {
    if (bodyOverflowStyle !== 'hidden' && !hidden && restrictScroll && trapFocus) {
      setBodyOverflowStyle(document.body.style.overflow);
      document.body.style.overflow = 'hidden';
    }
    return function () {
      if (bodyOverflowStyle !== 'hidden' && !hidden && restrictScroll && trapFocus) {
        document.body.style.overflow = bodyOverflowStyle;
      }
    };
  }, [bodyOverflowStyle, hidden, trapFocus, restrictScroll]);
  useEffect(function () {
    var timer = setTimeout(function () {
      if (!hidden && trapFocus && roots && roots[0]) {
        roots.forEach(makeNodeUnfocusable);
      }
    }, 0);
    return function () {
      // remove trap and restore ability to focus on the last root only
      if (roots && roots[0]) makeNodeFocusable(roots[roots.length - 1]);
      clearTimeout(timer);
    };
  }, [hidden, roots, trapFocus]);
  return /*#__PURE__*/React.createElement(RootsContext.Provider, {
    value: nextRoots
  }, /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    "aria-hidden": hidden
  }, rest), children));
};