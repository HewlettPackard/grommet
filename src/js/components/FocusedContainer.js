import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { findDOMNode } from 'react-dom';

import { getBodyChildElements, makeNodeFocusable, makeNodeUnfocusable } from '../utils';

const isNotAncestorOf = child => parent => !parent.contains(child);

export default class FocusedContainer extends Component {
  static defaultProps = {
    hidden: false,
    restrictScroll: false,
  }

  static propTypes = {
    hidden: PropTypes.bool,
    restrictScroll: PropTypes.bool,
  }

  componentDidMount() {
    const { hidden } = this.props;
    if (!hidden) {
      this.trapFocus();
    }
  }

  componentWillReceiveProps({ hidden }) {
    if (hidden !== this.props.hidden) {
      if (!hidden) {
        this.trapFocus();
      } else {
        this.removeTrap();
      }
    }
  }

  componentWillUnmount() {
    this.removeTrap();
  }

  removeTrap = () => {
    const { restrictScroll } = this.props;
    const child = findDOMNode(this.ref);
    getBodyChildElements()
      .filter(isNotAncestorOf(child))
      .forEach(makeNodeFocusable);
    if (restrictScroll) {
      document.body.style.overflow = 'scroll';
    }
  }

  trapFocus = () => {
    const { restrictScroll } = this.props;
    const child = findDOMNode(this.ref);
    getBodyChildElements()
      .filter(isNotAncestorOf(child))
      .forEach(makeNodeUnfocusable);

    if (restrictScroll) {
      document.body.style.overflow = 'hidden';
    }
  }

  render() {
    const { children, ...rest } = this.props;
    delete rest.restrictScroll;
    return (
      <div ref={(ref) => { this.ref = ref; }} onBlur={this.removeTrap} {...rest}>
        {children}
      </div>
    );
  }
}
