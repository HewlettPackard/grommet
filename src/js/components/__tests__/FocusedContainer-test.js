import React from 'react';
import 'jest-styled-components';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import FocusedContainer from '../FocusedContainer';

Enzyme.configure({ adapter: new Adapter() });

const createContainer = () => {
  document.body.appendChild(document.createElement('div'));
};

describe('FocusedContainer', () => {
  beforeEach(() => {
    // make sure to remove all body children
    document.body.innerHTML = '';
    createContainer();
  });

  test('mounts', () => {
    mount(
      <div id='focus-trap-test'><input id='test' /></div>, {
        attachTo: document.body.firstChild,
      }
    );
    const element = document.createElement('div');
    document.body.appendChild(element);
    const component = mount(
      <FocusedContainer id='container'>test focused container</FocusedContainer>, {
        attachTo: element,
      }
    );
    expect(component.getDOMNode()).toMatchSnapshot();
    expect(document.getElementById('focus-trap-test')).toMatchSnapshot();

    component.setProps({ hidden: true });

    expect(component.getDOMNode()).toMatchSnapshot();
    expect(document.getElementById('focus-trap-test')).toMatchSnapshot();

    component.setProps({ hidden: false });

    expect(component.getDOMNode()).toMatchSnapshot();
    expect(document.getElementById('focus-trap-test')).toMatchSnapshot();

    document.getElementById('test').focus();
  });

  test('restrict scroll', () => {
    const component = mount(
      <FocusedContainer id='container' restrictScroll={true}>
        test focused container
      </FocusedContainer>, {
        attachTo: document.body.firstChild,
      }
    );
    expect(component.getDOMNode()).toMatchSnapshot();
    expect(document.body.style.overflow).toMatchSnapshot();

    component.unmount();

    expect(component.getDOMNode()).toBeNull();
    expect(document.body.style.overflow).toMatchSnapshot();
  });

  test('blurs', () => {
    mount(
      <div id='focus-trap-test'><input id='test' /></div>, {
        attachTo: document.body.firstChild,
      }
    );
    const element = document.createElement('div');
    document.body.appendChild(element);
    const component = mount(
      <FocusedContainer id='container'>test focused container</FocusedContainer>, {
        attachTo: element,
      }
    );
    expect(component.getDOMNode()).toMatchSnapshot();
    expect(document.getElementById('focus-trap-test')).toMatchSnapshot();

    component.simulate('blur');

    expect(document.getElementById('focus-trap-test')).toMatchSnapshot();
  });
});
