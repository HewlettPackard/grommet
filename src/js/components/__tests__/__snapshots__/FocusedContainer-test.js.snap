// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusedContainer blurs 1`] = `
<div
  id="container"
>
  test focused container
</div>
`;

exports[`FocusedContainer blurs 2`] = `
<div
  id="focus-trap-test"
>
  <input
    id="test"
    tabindex="-1"
  />
</div>
`;

exports[`FocusedContainer blurs 3`] = `
<div
  id="focus-trap-test"
>
  <input
    id="test"
  />
</div>
`;

exports[`FocusedContainer mounts 1`] = `
<div
  id="container"
>
  test focused container
</div>
`;

exports[`FocusedContainer mounts 2`] = `
<div
  id="focus-trap-test"
>
  <input
    id="test"
    tabindex="-1"
  />
</div>
`;

exports[`FocusedContainer mounts 3`] = `
<div
  id="container"
  hidden=""
>
  test focused container
</div>
`;

exports[`FocusedContainer mounts 4`] = `
<div
  id="focus-trap-test"
>
  <input
    id="test"
  />
</div>
`;

exports[`FocusedContainer mounts 5`] = `
<div
  id="container"
>
  test focused container
</div>
`;

exports[`FocusedContainer mounts 6`] = `
<div
  id="focus-trap-test"
>
  <input
    id="test"
    tabindex="-1"
  />
</div>
`;

exports[`FocusedContainer restrict scroll 1`] = `
<div
  id="container"
>
  test focused container
</div>
`;

exports[`FocusedContainer restrict scroll 2`] = `"hidden"`;

exports[`FocusedContainer restrict scroll 3`] = `"scroll"`;
